### ISSUE TITLE
Upgrade to Kubernetes 1.26

### ISSUE THREAD

**Issue by dduportal** [jenkins-infra/helpdesk#3683]
Previous upgrade (1.25): https://github.com/jenkins-infra/helpdesk/issues/3582

Depreciation timelines for 1.25 (justifying the upgrade to 1.26):

- Digital Ocean: **Oct. 2023** - https://docs.digitalocean.com/products/kubernetes/details/supported-releases/
- Azure AKS: Dec. 2023 - https://learn.microsoft.com/en-us/azure/aks/supported-kubernetes-versions?tabs=azure-cli#aks-kubernetes-release-calendar
- AWS EKS: May 2024 - https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#kubernetes-release-calendar

----

Task list:

- [x] Upgrade kubectl within [docker-helmfile](https://github.com/jenkins-infra/docker-helmfile)
   - [x]  First upgrade the updatecli manifest <add link here>
   - [x] Then apply the updatecli PR that is produced by this change: <add link here>
   - [x] Then Apply the Kubernetes-Management PR that will use this new version - <add link here>

- [x] Upgrade DOKS (`doks` and `doks-public`) - see https://github.com/jenkins-infra/helpdesk/issues/3683#issuecomment-1780857936 below

- [x] Upgrade AWS EKS (`cik8s` and `eks-public`) - see https://github.com/jenkins-infra/helpdesk/issues/3683#issuecomment-1794913444
 
- [ ] Upgrade AKS (`privatek8s` and `publick8s`) - https://github.com/jenkins-infra/helpdesk/issues/3683#issuecomment-1797928819

**Comment by lemeurherve** [comment]
To keep in mind: https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke

**Comment by smerle33** [comment]
kubectl 1.26.9 is now used in our system (client side).

**Comment by dduportal** [comment]
Update:

- WiP on DigitalOcean clusters:
  - Changelog for Kube 1.26 being checked with a focus on https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.26.md#urgent-upgrade-notes 
    - No breaking changes for DO and we don't use directly the deprecated APIs (https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-26). Eventually some of the 3rd-party helm-chart we use do it, but the upgrade and an helmfile apply will fix it (if not already done with 1.25 server-side and 1.26 client-side)
  - DO has their own changelog for their managed components: https://docs.digitalocean.com/products/kubernetes/details/changelog/#1-26-x
    - All components are upgraded as expected, no major change catching attention (but it is managed so nothing to do really)
    - Note: Debian on the nodes will be upgraded featuring a kernel 6.x
  - Preparing upgrade of the cluster, the 2 at the same time (`doks-public` and `doks`)
    - Requires disabling DO from agent clouds in ci.jenkins.io at least 15 min before the operation
    - Need to announce operation (pick a date time, and then usual status/mailing/IRC, etc.)

TODO:

- EKS
- AKS

**Comment by dduportal** [comment]
Update on DigitalOcean upgrade:

- Announcement: Thursday 26 Oct. at noon UTC
  - [ ] status.jenkins.io
  - [ ] developer mailing list (as ci.jenkins.io need DOKS to be disabled for agents) + jenkins-infra mailing
  - [ ] IRC #jenkins-infra
- [ ] Disable DO Kubernetes Cloud agent in ci.jenkins.io (puppet hieradata PR)
- [ ] Disable kubernetes-management of both clusters (PR)
- [ ] Bump Kube version to 1.26 for doks-public
  - [ ] Terraform PR
  - [ ] Sanity check with a local helmfile diff
- [ ] Bump Kube version to 1.26 for doks
  - [ ] Terraform PR
  - [ ] Sanity check with a local helmfile diff
- [ ] Enable kubernetes-management of both clusters (PR)
- [ ] Enable DO Kubernetes Cloud agent in ci.jenkins.io (puppet hieradata PR)
- [ ] Validate agents can be spawned using a replay of the Infra/acceptance-tests and a simple scripted pipeline `node('doks && maven') { sh 'mvn -v' }`
- [ ] Close announcements

**Comment by dduportal** [comment]
EKS changelogs:

- [Reminder: Upgrade note from 1.25 to 1.26 for Kubernetes itself](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.26.md#no-really-you-must-read-this-before-you-upgrade-1)
- [EKS 1.26 Release Notes](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions-standard.html#kubernetes-1.26)
- [A testimony of a Terraform-managed EKS upgrade from 1.25 to 1.26](https://marcincuber.medium.com/amazon-eks-upgrade-journey-from-1-25-to-1-26-electrifying-79b287084eef)


TL;DR:

- We have to check for the VPC CNI plugins to be on version `1.12` or later **before** the upgrade
- No problem on the containerd version (removed CRI v1alpha2) as we are using Amazon default AMI images
- With 1.26, EKS is finally having the same feature set for CSI `fsGroup` option. We don't really care as only our ACP services uses persistent volume in EKS, but worth having in mind

**Comment by dduportal** [comment]
Update on AWS EKS upgrade: <WIP>

**Comment by dduportal** [comment]
Update: AKS Upgrade plan

- [ ] Check changelog:

Current [changelog](https://github.com/Azure/AKS/blob/master/CHANGELOG.md) notable elements for Kubernetes 1.26:

- Some AKS labels are being deprecated with the Kubernetes 1.26 release. Update your AKS labels to the recommended substitutions. See more information on label deprecations and how to update your labels in the Use labels in an AKS cluster documentation. beta.kubernetes.io/arch= and beta.kubernetes.io/os= are still applied by kubelet in kubernetes code.
- HostProcess Containers will be GA
- Two in-tree driver persistent volumes won't be supported in AKS : kubernetes.io/azure-disk, kubernetes.io/azure-file.
- All AKS clusters on version 1.26+ will use the latest coreDNS version v1.10.1..
  - For all AKS clusters on version 1.26+, coreDNS health plugin will use lameduck 5s to minimizes DNS resolution failures during coreDNS pod restart or deployment rollout.
  - For all AKS clusters on version 1.26+, coreDNS will use ttl 30 as default TTL for DNS records.
- During cluster upgrade to v1.26.0 or a later version, disk PV node affinity check will cause the upgrade to fail if there are disk PVs still using deprecated labels: failure-domain.beta.kubernetes.io/zone and failure-domain.beta.kubernetes.io/region
  - Enable failure-domain.beta.kubernetes.io labels on K8S 1.26+ nodes by default to resolve issue with in tree CSI drivers. Will be removed from K8S 1.28
  
**privatek8s**

- Announcement: Tuesday 07 Nov. 2023 at 08:00 UTC
  - [ ] Open datetime announcement on status.jenkins.io - <add link here>
  - [ ] jenkins-infra mailing - 
  - [ ] IRC #jenkins-infra - 
- [ ] Disable kubernetes management for privatek8s cluster (remove it from Jenkinsfile_k8s)
- [ ] Terraform PR to upgrade cluster to 1.26 but NO merge
- [ ] Disable terraform azure job (⚠️ As the upgrade will put infra.ci out of service so need to be run manually)
- [ ] Make sure local (on your machine) repository is up to date and you are logged-in: run a `terraform plan`
- [ ] Merge Terraform PR even without checks (as infra.ci restart will re-enable job)
- [ ] Perform upgrade from local machine
- [ ] Check that helmfile diff still works and shows no change and no warning
- [ ] Wait for infra.ci to be up again
- [ ] Enable kubernetes management for privatek8s cluster and check it runs successfully
- [ ] Check terraform job and eventually trigger it for success with no changes
- [ ] Check for release.ci to be up and running: try to allocate agents on release.ci with a job replay (dummy linux and a dummy window)

**publick8s**

- Announcement: Tuesday 07 Nov. 2023 at 13:00 UTC
  - [ ] Open datetime announcement on status.jenkins.io - <add link here>
  - [ ] jenkins-dev and jenkins-infra mailing - 
  - [ ] IRC #jenkins-infra - 
- [ ] Disable kubernetes management for publick8s cluster (remove it from Jenkinsfile_k8s)
- [ ] Terraform PR to upgrade cluster to 1.26 
- [ ] Merge Terraform PR and let the upgrade happen
- [ ] Check that helmfile diff still works and shows no change and no warning
- [ ] Enable kubernetes management for publick8s cluster and check it runs successfully
- [ ] Check for services healthiness (pagerduty, datadog, manual control, logging in on ci.jenkins.io, etc.)

**Comment by dduportal** [comment]
**publick8s**

- Announcement: Tuesday 07 Nov. 2023 at 13:00 UTC
  - [ ] Open datetime announcement on status.jenkins.io - <add link here>
  - [ ] jenkins-dev and jenkins-infra mailing - 
  - [ ] IRC #jenkins-infra - 
- [ ] Disable kubernetes management for publick8s cluster (remove it from Jenkinsfile_k8s)
- [ ] Terraform PR to upgrade cluster to 1.26 
- [ ] Merge Terraform PR and let the upgrade happen
- [ ] Check that helmfile diff still works and shows no change and no warning
- [ ] Enable kubernetes management for publick8s cluster and check it runs successfully
- [ ] Check for services healthiness (pagerduty, datadog, manual control, logging in on ci.jenkins.io, etc.)

**Comment by dduportal** [comment]
# Operation on publick8s (wip)

## Chart upgrades (falco and Nginx Ingress)

- [ ] Fix Falco crashes on arm64 nodes by upgrading it to 3.x
  - Rationale: Falco needed a (major) upgrade as its daemonset images are crashing on the latest arm64 nodes of AKS 1.25.
  - PR: https://github.com/jenkins-infra/kubernetes-management/pull/4619
- [ ] Upgrade Nginx ingress to latest patch version with revamped NetworkSecurities
  - Rationale: impact on updated/added ingresses can be detected long after the upgrade. Doing this *before* a full kube upgrade allows detecting issues immediately
  - PR: https://github.com/jenkins-infra/kubernetes-management/pull/4622

## Public IP 

## Kubernetes 1.26 upgrade

**Comment by dduportal** [comment]
Mandatory logo

![kube-1 26-logo](https://github.com/jenkins-infra/helpdesk/assets/1522731/8b440be3-c16a-489d-bd47-82df3c4cf678)


@smerle33 I'll let you close this issue with a mandatory gif or image ;)

**Comment by smerle33** [comment]
![image](https://github.com/jenkins-infra/helpdesk/assets/95630726/3ce549d8-5bb6-430d-aa83-2f61d3863246)
